File: src/main/java/by/auction/service/ProductService.java
package by.auction.service;

import by.auction.entity.Product;

import java.util.List;
import java.util.Optional;

/**
 * Service for data management in a table "products"
 */
public interface ProductService {

    /**
     * Find all products
     * @return list of found products
     */
    List<Product> findAll();

    /**
     * Find product by ID
     * @param id
     * @return Optional of found product
     */
    Optional<Product> findById(Long id);

    /**
     * Save product in DB
     * @param product
     * @return saved product
     */
    Product save(final Product product);

    /**
     * Delete product by ID
     * @param id
     */
    void deleteById(Long id);

}

File: auction-client-angular/src/app/services/product/product.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs/Observable';
import {Product} from '../../models/product';
import {ConfigService} from "../config/config.service";

/**
 * Service to access products data
 */
@Injectable()
export class ProductService {

  /**
   * Constructor for product service
   * @param {HttpClient} http
   * @param {ConfigService} config
   */
  constructor(private http: HttpClient,
              private config: ConfigService) {
  }

  /**
   * Get api uri
   * @returns {string}
   */
  uri() {
    return this.config.getApiHref() + 'products';
  }

  /**
   * Find all products
   * @returns {Observable<Product[]>}
   */
  findAll(): Observable<Product[]> {
    return this.http.get<Product[]>(this.uri());
  }

  /**
   * Find product by id
   * @param {number} id
   * @returns {Observable<Product>}
   */
  findById(id: number): Observable<Product> {
    return this.http.get<Product>(this.uri() + '/' + id);
  }

  /**
   * Save product
   * @param {Product} product
   * @returns {Observable<Product>}
   */
  save(product: Product): Observable<Product> {
    return this.http.post<Product>(this.uri(), product);
  }

  /**
   * Delete product by id
   * @param {number} id
   * @returns {Observable<Object>}
   */
  deleteById(id: number) {
    return this.http.delete(this.uri() + '/' + id);
  }
}
